### PR TITLE
enhance: Skip timestamp allocation when search/query consistency level is eventually

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #29772 
1. Skip timestamp allocation when search/query consistency level is eventually.

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by samhuang-z** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/proxy/task_query.go
Comment (longjiquan) [inline internal/proxy/task_query.go:259]: I think this check should be moved above, which may also skip to get collection schema from rootcoord.

File: internal/proxy/task_scheduler.go
Comment (longjiquan) [inline internal/proxy/task_scheduler.go:192]: The logic is a little bit duplicate with below `SetID` & `addUnissuedTask`. Maybe we can consider only spliting how we allocate the timestamp.

File: internal/proxy/task_search.go
Comment (longjiquan) [inline internal/proxy/task_search.go:255]: Ditto.

File: internal/proxy/util.go
Comment (longjiquan) [inline internal/proxy/util.go:782]: Quick question, what time will `tMax` be `0`?

**Review by longjiquan — CHANGES_REQUESTED** [review]

**Comment by longjiquan** [comment]
LGTM generally for the code changes.

File: internal/proxy/task_scheduler.go
Comment (longjiquan) [inline internal/proxy/task_scheduler.go:184]: Why we don't `SetTs` for tasks which can skip allocating timestamp?

**Review by longjiquan — CHANGES_REQUESTED** [review]

**Comment by bigsheeper** [comment]
https://github.com/milvus-io/milvus/blob/930735c46cab6e0cefd5be2d8707ebcb613d1eed/internal/metastore/catalog.go#L20-L25

File: internal/proxy/task_scheduler.go
Comment (xiaocai2333) [reply internal/proxy/task_scheduler.go:184]: This timestamp will be set to MVCCTS. In querynode, there is logic that if MVCCTS is 0, it will be assigned the value tsafe. If ts is set here, it may affect the query.

https://github.com/czs007/milvus/blob/b9d76f77d19810f8f7c8f760745bdb5c6495c61f/internal/querynodev2/delegator/delegator.go#L290-L292

Comment (xiaocai2333) [reply internal/proxy/task_scheduler.go:184]: https://github.com/czs007/milvus/blob/b9d76f77d19810f8f7c8f760745bdb5c6495c61f/internal/querynodev2/delegator/delegator.go#L290-L292

**Comment by xiaocai2333** [comment]
https://github.com/czs007/milvus/blob/b9d76f77d19810f8f7c8f760745bdb5c6495c61f/internal/querynodev2/delegator/delegator.go#L290-L292

**Review by yiwangdr — COMMENTED** [review]
Do we have integration test / e2e test that can cover the changes in this PR?

File: internal/proxy/task_query.go
Comment (xiaocai2333) [reply internal/proxy/task_query.go:266]: After discussion, Bound is also an estimated time, so here it is followed by eventually.
But it occurred to me that regardless of consistency, we should decide whether to skip allocating timestamps after trying to allocate timestamps and failing.

I will update it.

**Comment by xiaocai2333** [comment]
> Do we have integration test / e2e test that can cover the changes in this PR?

This kind of test should belong to chaos test.

File: internal/proxy/task_query.go
Comment (yiwangdr) [reply internal/proxy/task_query.go:266]: Thanks for being thoughtful. 
Please add your thoughts as comments above the code.

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
@czs007 @xiaofan-luan 
There are two options here:
1. Regardless of whether RootCoord crashes, search/query requests with consistency level eventually/bound skip allocating timestamps. So, **when RootCoord does not crash, these requests will not try to allocate timestamps**.
2. First try to allocate the timestamp, if the allocation fails due to RootCoord crash, then skip timestamp allocation for search/query requests with consistency level eventually/bound. However, the **delay** of these search/query requests will be **greater than 5s**(try to allocate timestamps, but fails).

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by yiwangdr** [comment]
@czs007 could you help to merge this PR if it looks good to you?

File: internal/proxy/util.go
Comment (czs007) [inline internal/proxy/util.go:783]: should remove this

File: internal/proxy/task_query.go
Comment (czs007) [inline internal/proxy/task_query.go:249]: cache schema into task object

Comment (xiaocai2333) [reply internal/proxy/task_query.go:251]: `collectionInfo` only contains the consistency level, not the schema.

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/proxy/task_query.go
Comment (xiaofan-luan) [inline internal/proxy/task_query.go:267]: change to consistencyLevel != commonpb.ConsistencyLevel_Strong

**Comment by bigsheeper** [comment]
related: https://github.com/milvus-io/milvus/issues/30299

File: internal/proxy/meta_cache.go
Comment (czs007) [inline internal/proxy/meta_cache.go:998]: change to AllocID

File: internal/proxy/meta_cache.go
Comment (czs007) [inline internal/proxy/meta_cache.go:998]: use another mutext?

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/proxy/task_scheduler.go
Comment (xiaofan-luan) [inline internal/proxy/task_scheduler.go:177]: does that mean we need a unique messageID?

If that's the case why don't we just use a local tso instead of using global ids?

File: internal/proxy/task_scheduler.go
Comment (xiaofan-luan) [inline internal/proxy/task_scheduler.go:176]: can we skip alloc timestamp for other operations?

Comment (xiaocai2333) [reply internal/proxy/task_scheduler.go:176]: No, for CreateCollection,  cannot skip alloc timestamp.

File: internal/proxy/task_scheduler.go
Comment (xiaocai2333) [reply internal/proxy/task_scheduler.go:177]: yes, we need a unique messageID, so we cache a large number of global ids here. If you use local tso, duplicate ids may appear.

**Pull request merged**
